[SelectionDAG] Legalize <1 x T> vector types for atomic load

jofrn opened the pull request:
`load atomic <1 x T>` is not valid. This change legalizes vector types of atomic load via scalarization in SelectionDAG so that it can, for example, translate from `v1i32` to `i32`.

jofrn committed:
[SelectionDAG] Legalize vector types for atomic load

Scalarize vector of atomic load in SelectionDAG.

arsenm commented on llvm/test/CodeGen/Generic/atomic-scalarization.ll:
```diff
@@ -0,0 +1,17 @@
+; RUN: llc %s --print-after-isel --disable-verify 2>&1 | FileCheck %s
```
Don't use print-after-isel, or -disable-verify. Check the final isa output (or -stop-after to check real mir), and move to test/CodeGen/X86

arsenm replied:
Maybe -stop-after=finalize-isel

jofrn replied:
Why `-stop-after=finalize-isel` yet not `--print-after-isel`?

arsenm commented on llvm/test/CodeGen/Generic/atomic-scalarization.ll:
```diff
@@ -0,0 +1,17 @@
+  %2 = load atomic <1 x i32>, ptr %1 acquire, align 4
+  %3 = extractelement <1 x i32> %2, i32 0
+  ret i32 %3
+}
```
Test an fp case too

jofrn committed:
Moved test and checking mir after last X86 pass

arsenm commented on llvm/test/CodeGen/X86/atomic-scalarization.ll:
```diff
@@ -0,0 +1,16 @@
+; RUN: llc %s --mtriple=x86_64 -print-after=unpack-mi-bundles -disable-verify 2>&1 | FileCheck %s
```
Use -stop-after, do not rely on -print-after output. It is NOT real mir. Also don't use disable-verify. Also just use finalize-isel

jofrn replied:
Ah, will use -stop-after. thanks.. Atomic load with vector is not legal as per IR/Verifier, so we need disable-verify as far as I can tell.

arsenm replied:
Then this shouldn't be handled at all

jofrn committed:
Autogenerate test.

arsenm reviewed (approved):
Do stores also need to be handled?

arsenm reviewed (changes requested):
No disable-verify

jofrn committed:
[Verifier] Allow vector type in atomic load and store

Vector types on atomics are assumed to be invalid by the verifier. However,
this type can be valid if it is lowered by codegen.

jofrn committed:
Update Assembler/atomic test

jofrn commented:
| Do stores also need to be handled?
No, they do not.
| No disable-verify
Rebased this on top of https://github.com/llvm/llvm-project/pull/117625.

arsenm commented:
> | Do stores also need to be handled? No, they do not.

They do, but it can be a separate pr

arsenm commented on llvm/test/CodeGen/X86/atomic-scalarization.ll:
```diff
@@ -0,0 +1,21 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 5
+; RUN: llc %s --mtriple=x86_64 -o - | FileCheck %s
+
```
test name could be more specific, and preferably merged with an existing atomic test. With more run lines covering the different legal types.

arsenm reviewed (commented):
codegen parts lgtm, but should avoid manually stacking PRs (which really doesn't work). It's much harder to review, if you want to stack PRs use graphite or spr

jofrn committed:
Renamed to atomic-vector

jofrn committed:
Add to preexisting test

jofrn commented on llvm/test/CodeGen/X86/atomic-load-store.ll:
```diff
@@ -1,12 +1,18 @@
 ; RUN: llc < %s -mtriple=x86_64-apple-macosx10.7.0 -verify-machineinstrs | FileCheck %s
 ; RUN: llc < %s -mtriple=x86_64-apple-macosx10.7.0 -verify-machineinstrs -O0 | FileCheck %s
+; RUN: llc %s --mtriple=x86_64 -o - | FileCheck %s --check-prefix=CHECK0
 
```
This seems like a good preexisting test to add the <1 x n> cases to. It has a specific triple `apple`, and when generalizing it, the difference in the output tests is not much, and so it duplicates quite a bit. However, I do notice that bfloat passes only when `--mtriple=x86_64`, so we may want to add the additional run line.

arsenm replied:
Just copy the mtriple that's already here, the OS should not be significant for the purpose here

arsenm reviewed (commented):
This patch will only cover 1 x vectors. The patch is fine as-is, but the message should change to reflect this. An additional step is required to handle any other vector size

jofrn committed:
Rename tests

arsenm commented on llvm/lib/CodeGen/SelectionDAG/LegalizeVectorTypes.cpp:
```diff
@@ -451,6 +454,19 @@ SDValue DAGTypeLegalizer::ScalarizeVecRes_INSERT_VECTOR_ELT(SDNode *N) {
 }
 
+SDValue DAGTypeLegalizer::ScalarizeVecRes_ATOMIC_LOAD(AtomicSDNode *N) {
+
```
```suggestion
```

arsenm commented:
> This patch will only cover 1 x vectors. The patch is fine as-is, but the message should change to reflect this. An additional step is required to handle any other vector size

Still only handles 1 x vectors

jofrn committed:
Fix whitespace

RKSimon commented on llvm/test/Verifier/atomics.ll:
```diff
@@ -1,14 +1,15 @@
+define %ty @bar(ptr %P) {
+  %v = load atomic %ty, ptr %P unordered, align 8
+  ret %ty %v
 }
```
please can you add <1 x i64> load/store test coverage some place

jofrn replied:
https://github.com/llvm/llvm-project/blob/fb8e577977fd3c1468d573f8597fb2896e4e7f72/llvm/test/CodeGen/X86/atomic-load-store.ll#L113 Just added it here https://github.com/llvm/llvm-project/pull/120316/commits.

arsenm replied:
Also pointers (including the 32-bit addrspace on x86-64 case which is some high number)

arsenm reviewed (approved):
DAG change LGTM on top of the additional tests in the other PR
